vkubelet/podcontroller.go

```diff
@@ -0,0 +1,380 @@
+
+	// Convert the namespace/name string into a distinct namespace and name.
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// Log the error as a warning, but do not requeue the key as it is invalid.
+		log.G(ctx).Warn(pkgerrors.Wrapf(err, "invalid resource key: %q", key))
+		return nil
+	}
+
+	// Get the Pod resource with this namespace/name.
+	pod, err := pc.podsLister.Pods(namespace).Get(name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			// We've failed to fetch the pod from the lister, but the error is not a 404.
+			// Hence, we add the key back to the work queue so we can retry processing it later.
+			err := pkgerrors.Wrapf(err, "failed to fetch pod with key %q from lister", key)
+			span.SetStatus(ocstatus.FromError(err))
+			return err
+		}
+		// At this point we know the Pod resource doesn't exist, which most probably means it was deleted.
+		// Hence, we must delete it from the provider if it still exists there.
+		if err := pc.server.deletePod(ctx, namespace, name); err != nil {
+			err := pkgerrors.Wrapf(err, "failed to delete pod %q in the provider", loggablePodNameFromCoordinates(namespace, name))
+			span.SetStatus(ocstatus.FromError(err))
+			return err
+		}
+		return nil
+	}
+	// At this point we know the Pod resource has either been created or updated (which includes being marked for deletion).
+	return pc.syncPodInProvider(ctx, pod)
+}
+
+// syncPodInProvider tries and reconciles the state of a pod by comparing its Kubernetes representation and the provider's representation.
+func (pc *PodController) syncPodInProvider(ctx context.Context, pod *corev1.Pod) error {
+	// Create a span for the current call.
+	ctx, span := trace.StartSpan(ctx, "syncPodInProvider")
+	defer span.End()
+
+	// Add the pod's attributes to the current span.
+	addPodAttributes(span, pod)
+
+	// Check whether the pod has been marked for deletion.
+	// If it does, delete it in the provider.
+	if pod.DeletionTimestamp != nil {
+		// Delete the pod.
+		if err := pc.server.deletePod(ctx, pod.Namespace, pod.Name); err != nil {
+			err := pkgerrors.Wrapf(err, "failed to delete pod %q in the provider", loggablePodName(pod))
+			span.SetStatus(ocstatus.FromError(err))
+			return err
+		}
+		return nil
+	}
+
+	// Ignore the pod if it is in the "Failed" state.
+	if pod.Status.Phase == corev1.PodFailed {
+		log.G(ctx).Warnf("skipping sync of pod %q in %q phase", loggablePodName(pod), pod.Status.Phase)
+		return nil
+	}
+
+	// Create or update the pod in the provider.
+	if err := pc.server.createOrUpdatePod(ctx, pod); err != nil {
+		err := pkgerrors.Wrapf(err, "failed to sync pod %q in the provider", loggablePodName(pod))
+		span.SetStatus(ocstatus.FromError(err))
+		return err
+	}
+	return nil
+}
+
+// deleteDanglingPods checks whether the provider knows about any pods which Kubernetes doesn't know about, and deletes them.
+func (pc *PodController) deleteDanglingPods() {
```

Comment:
Curious what others may think... do we need to stop the world for this collection?

Reply:
I strongly believe we should start the controller only after this initial reconciliation. FWIW, VK, as is in the master branch, stops the world for [something that is almost identical to this](https://github.com/virtual-kubelet/virtual-kubelet/blob/master/vkubelet/pod.go#L326-L327).

Reply:
Ah yes. Was thinking this was being run periodically but I guess it's not. This is fine.